src/Microsoft.DotNet.Build.Tasks/NuGetPush.cs

```diff
@@ -0,0 +1,140 @@
+    /// https://github.com/NuGet/Home/issues/5224 requests retrying on hangs in NuGet.exe.
+    /// </summary>
+    public class NuGetPush : Task, ICancelableTask
+    {
+        [Required]
+        public string NuGetExePath { get; set; }
+
+        [Required]
+        public ITaskItem[] Inputs { get; set; }
+
+        [Required]
+        public string Source { get; set; }
+
+        public string ApiKey { get; set; }
+
+        public string Verbosity { get; set; } = "Detailed";
+
+        public int TimeoutPerPushInSeconds { get; set; } = TimeSpan.FromMinutes(5).Seconds;
+
+        public int RetriesPerPush { get; set; } = 5;
+
+        private bool _cancelRemainingPushes;
+
+        public void Cancel()
+        {
+            _cancelRemainingPushes = true;
+        }
+
+        public override bool Execute()
+        {
+            foreach (ITaskItem input in Inputs)
+            {
+                for (int i = 0; i < RetriesPerPush; i++)
+                {
+                    if (_cancelRemainingPushes)
+                    {
+                        break;
+                    }
+
+                    var args = new List<string>
+                    {
+                        "push", input.ItemSpec,
+                        "-Source", Source,
+                        "-Timeout", TimeoutPerPushInSeconds.ToString(),
+                        "-Verbosity", Verbosity
+                    };
+
+                    if (!string.IsNullOrEmpty(ApiKey))
+                    {
+                        args.Add("-ApiKey");
+                        args.Add(ApiKey);
+                    }
+
+                    string attemptMessage = $"(attempt {i + 1}/{RetriesPerPush})";
+
+                    Log.LogMessage(
+                        MessageImportance.Normal,
+                        $"Pushing '{input.ItemSpec}' {attemptMessage} ...");
+
+                    int exitCode = RunProcess(NuGetExePath, args);
+                    if (exitCode == 0)
+                    {
+                        Log.LogMessage(
+                            MessageImportance.Normal,
+                            $"Pushing '{input.ItemSpec}' {attemptMessage} Done.");
+
+                        break;
+                    }
+
+                    string message =
+                        $"Pushing '{input.ItemSpec}' {attemptMessage} FAILED: " +
+                        $"exit code {exitCode}";
+
+                    if (i + 1 == RetriesPerPush)
+                    {
+                        Log.LogError(message);
+                    }
+                    else
+                    {
+                        Log.LogMessage(message);
+                    }
+                }
+            }
+            return !Log.HasLoggedErrors;
+        }
+
+        private int RunProcess(string exe, IEnumerable<string> args)
+        {
+            using (var pushProcess = new Process
+            {
+                StartInfo = new ProcessStartInfo(NuGetExePath, string.Join(" ", args))
```

Comment:
> While that is reasonable we already do have a number of nuget dependencies in our tasks.

To me this seems like a reason to avoid making push also depend on NuGet packages. If it does, it's difficult to change the NuGet used to push because that changes the NuGet used for every other task. Using the latest supported NuGet.exe to push is typically MyGet's first suggestion when we hit flakiness.

For the CoreFX package set, this took 5.8 min for lib packages and 6.2 min for symbol packages, less total than the last official build at 5.4 min and 11.7 min. I'll measure this for CoreCLR, too.